EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,263 @@
+const typedData = [
+  {
+    'type': 'string',
+    'name': 'message',
+    'value': 'Hi, Alice!',
+  },
+  {
+    'type': 'uint',
+    'name': 'value',
+    'value': 42,
+  },
+];
+const signature = await web3.eth.signTypedData(signerAddress, typedData);
+```
+
+```javascript
+// Signer code JS example
+import * as _ from 'lodash';
+import * as ethAbi from 'ethereumjs-abi';
+
+const schema = _.map(typedData, entry => `${entry.type} ${entry.name}`).join(',');
+const schemaHash = ethAbi.soliditySHA3(['string'], schema);
+const data = _.map(typedData, 'value');
+const types = _.map(typedData, 'type');
+const hash = ethAbi.soliditySHA3(
+  ['bytes32', ...types],
+  [schemaHash, ...data],
+);
+```
+
+```solidity
+// Solidity example
+string message = 'Hi, Alice!';
+uint value = 42;
+bytes32 schemaHash = keccak256('string message', 'uint value'); // Probably hardcoded
+const hash = keccak256(schemaHash, message, value);
+address recoveredSignerAddress = ecrecover(hash, v, r, s);
+```
+
+Signature will be returned in the same format as `eth_sign` (the ecSignature params concatenated together (r + s + v) and hex encoded.
+
+## Rationale
+
+Signing support in Ethereum clients is most useful for off-chain transactions. On-chain transactions in Ethereum have well-defined typed schemas (gas uint, gasPrice uint, etc) which allows signer UIs to show useful information about transactions to the user. We need the same for off-chain transactions.
+
+The Solidity keccak256 function accepts an array of typed arguments and motivated the current design.
+
+The main and the most important part of this change is to give Signer UIs a chance to give a user enough information to be able to verify what is being signed even when using untrusted DApps. Since the DApp does not get to provide the signer with a hard-to-read hash but must supply the exact values (in plain-text) that the user will be signing, the dangers of signing a malicious action (transaction or otherwise) are greatly reduced.
+
+The schema information must be signed together with the data, so that the verifying party can be confident, that the user was shown the data with the corresponding type and names supplied by the DApp.
+
+This approach is even possible to implement on hardware wallets with small screens (Ledger Nano S, Trezor). Users will be shown one line at a time (as currently with transactions).
+
+
+## Backwards Compatibility
+
+This EIP does not break backwards compatibility.
+
+`eth_sign` and `personal_sign` will not be removed for backwards compatibility reasons even though the `eth_signTypedData` is a superset of their functionality. Current off-chain protocols will however need to modify the verifying code in their smart contracts to include the schema.
+
+Example of signing a simple string using `eth_signTypedData`
+
+```javascript
+const signature = await web3.eth.signTypedData([
+  {
+    'type': 'string',
+    'name': 'message',
+    'value': 'Hi, Alice!',
+  },
+]);
```

Comment:
How does the signer know which key to use?

Reply:
According to [line 119](https://github.com/ethereum/EIPs/pull/712/files/6be36e285e3776bf44a6a5efcba42706b72f925c#diff-8e6779bfd60c2dfb3dae7875b728decaR119), the first argument is the signer address. (note that in the actual jsonrpc call, the address is the second argument, just like `personal_sign`.

```javascript
const signature = await web3.eth.signTypedData(
  '0x5409ed021d9299bf6814279a6a1411a7e866a631',
  [{
    type: 'string',
    name: 'message',
    value: 'Hi, Alice!',
  }]
);
```

Reply:
@pirapira Sorry, can you clarify? Each entry has three constant keys: `name`, `value` and `type`

Reply:
@petejkim [Here](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign) the address is the first argument.
Also I've checked in my Metamask and it's the first argument

Reply:
@LogvinovLeon I meant private key.   In this example, the first argument of `signTypedData` is not an address.  It is an array.

Reply:
Fixed

Reply:
I feel relieved I was actually making sense.